modules/audio/src/main/native/android/dalvik/DalvikMusic.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2020 Gluon
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.helloandroid;
+
+import android.media.MediaPlayer;
+import android.media.MediaPlayer.OnCompletionListener;
+
+/**
+ * @author Almas Baimagambetov (almaslvl@gmail.com)
+ */
+public class DalvikMusic implements DalvikAudio {
+
+    private boolean isDisposed = false;
+    private MediaPlayer mediaPlayer;
+
+    public DalvikMusic(MediaPlayer mediaPlayer) {
+        this.mediaPlayer = mediaPlayer;
+    }
+
+    @Override
+    public void setLooping(boolean looping) {
+        mediaPlayer.setLooping(looping);
+    }
+
+    @Override
+    public void setVolume(double volume) {
+        mediaPlayer.setVolume((float)volume, (float)volume);
+    }
+
+    @Override
+    public void setOnFinished(final Runnable action) {
+        mediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
+            @Override
+            public void onCompletion(MediaPlayer mp) {
+                action.run();
+            }
+        });
+    }
+
+    @Override
+    public void play() {
+        mediaPlayer.start();
```

Comment:
Don't you need to listen to `setOnPreparedListener()` before you can safely call `start`?

Reply:
`prepare()` is called synchronously in `DalvikAudioService` before passing the media player to `DalvikMusic`, so I think that's fine? As for async stuff, I think it's better to let the user figure out the thread on which to load.

Reply:
For the VideoService, we use `mediaPlayer.prepareAsync()`, and then we listen to `onPrepared`, so `play()` can be called only once the media player is ready. For big files or Internet URLs I guess this makes sense, as it won't block the UI. 
This might also need the Media.Status enum (with a callback user will track media status to enable or disable buttons, for instance).